rfcs/2022-04-09-accessibility.md

```diff
@@ -0,0 +1,606 @@
+
+- People with low vision who rely on screen readers need to know when the modal opens.
+- People with low vision who cannot properly see the focus ring can get lost if focus isn't contained in the modal, this can be confusing and disorienting.
+- People with mobility issues using keyboard could tab out of the modal and have a hard time getting back into the modal.
+
+#### Best practices for focus traps: <!-- omit in toc -->
+
+- Ensure users don't get permanently trapped in the focus trap, If the user can open the modal with keyboard, they should also have the ability to exit the modal with keyboard.
+- The focus should loop when reaching the last or first element inside modal.
+- Convey the intent to screen reader users that there is a context switch, and they are now in a modal or in a focus trapped state.
+
+
+https://user-images.githubusercontent.com/35374649/170202482-b9af775d-4622-4b88-bda5-a4fa91395764.mov
+
+
+#### Platform Specific Implementation <!-- omit in toc -->
+
+**Web:**
+
+- [HTML `inert` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/inert)
+- [ReactAria FocusScope](https://react-spectrum.adobe.com/react-aria/FocusScope.html)
+
+**Inert**
+
+> [Demo](https://lx5rik.csb.app/)
+
+In the [HTML spec](https://github.com/WICG/inert/blob/main/explainer.md) there's a new proposal/property `inert`, when present, makes the browser "ignore" user input events for the element, including focus events and events from assistive technologies. This can be used for focus trapping.
+
+Downsides of inert:
+
+- Browser [support is very bad](https://caniuse.com/?search=inert), only works on modern browsers under experimental flag
+- Polyfill is [expensive performance-wise](https://github.com/WICG/inert#performance-and-gotchas)
+  Although the browser support for it is very bad, and we will need the [polyfill](https://github.com/WICG/inert).
+- Won't handle other edge cases like focus restoration.
+
+**ReactAria**
+
+> [Demo](https://codesandbox.io/s/blade-a11y-reactaria-focustrap-6z537u?file=/src/App.tsx)
+
+ReactAria's `FocusScope` component is a fully fledged solution with other edge cases handled for us.
+
+It also handles:
+
+- Focus restoration to trigger
+- Auto focusing to children elements
+
+Considering the `inert` attribute is not supported in major browsers & we will have to write custom logic to handle other edge cases, it's better to use the ReactAria's component in this case.
+
+**Native:**
+
+To simulate `inert` behaviour in react-native we have platform specific accessibility props:
+
+- For android: [`importantForAccessibility`](https://reactnative.dev/docs/accessibility#importantforaccessibility-android)
+- For iOS: [`accessibilityViewIsModal`](https://reactnative.dev/docs/accessibility#accessibilityviewismodal-ios) & [`accessibilityElementsHidden`](https://reactnative.dev/docs/accessibility#accessibilityelementshidden-ios)
+
```

Comment:
once you're done with all the evaluations, let's put a table for a summary of what to use for what platform for a particular a11y thing. it'll basically a snapshot of the information you've already put above. for eg: we'll use reac-aria for focus trap on web and on native we'll use the platform specific attributes, etc.